features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/metric/SessionEndedMetric.kt

```diff
@@ -0,0 +1,221 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.rum.internal.metric
+
+import com.datadog.android.core.metrics.PerformanceMetric
+import com.datadog.android.rum.internal.domain.scope.RumViewManagerScope
+import com.datadog.android.rum.model.ViewEvent
+
+/**
+ * Metric for rum session ended event.
+ */
+@Suppress("TooManyFunctions")
+internal class SessionEndedMetric(
+    private val sessionId: String,
+    private val precondition: String
+) {
+
+    private val trackedViewMap by lazy { mutableMapOf<String, TrackedView>() }
+
+    private val trackedErrorMap by lazy { mutableMapOf<String, Int>() }
+
+    private var firstTrackedView: TrackedView? = null
+    private var lastTrackedView: TrackedView? = null
+    private var wasStopped: Boolean = false
+
+    /**
+     * Called on view tracked event, return true if the view is recorded by the metric, false otherwise.
+     */
+    fun onViewTracked(rumViewEvent: ViewEvent): Boolean {
+        if (rumViewEvent.session.id != sessionId) {
+            return false
+        }
+        val trackedView = trackedViewMap[rumViewEvent.view.id] ?: TrackedView(
+            viewUrl = rumViewEvent.view.url,
+            startMs = rumViewEvent.date,
+            durationNs = rumViewEvent.view.timeSpent
+        )
+
+        trackedViewMap[rumViewEvent.view.id] = trackedView.copy(durationNs = rumViewEvent.view.timeSpent)
```

Comment:
nit: in case the view is tracked for the first time we are creating 2 objects here instead of one. This maybe could be optimised but is not that bad.